### ISSUE TITLE
Performance regressions in `pl.LazyFrame.drop()` and `pl.LazyFrame.collect()` introduced in `polars==0.19.2` (~5-10x slower)

### ISSUE THREAD

**Issue by jakob-keller** [pola-rs/polars#10939]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

columns = tuple(f"column_{i}" for i in range(100))
ldf = pl.LazyFrame({column: tuple(range(10)) for column in columns})

# isolated .drop() executes 5-10x slower
ldf.drop(columns[:len(columns) // 2])

# combined .drop().collect() also executes 5-10x slower
ldf.drop(columns[:len(columns) // 2]).collect()
```

May be timed by calling:
`python -m timeit -s "import polars as pl; columns = tuple(f'column_{i}' for i in range(100)); ldf = pl.LazyFrame({column: tuple(range(10)) for column in columns})" "ldf.drop(columns[:len(columns) // 2])"`

`python -m timeit -s "import polars as pl; columns = tuple(f'column_{i}' for i in range(100)); ldf = pl.LazyFrame({column: tuple(range(10)) for column in columns})" "ldf.drop(columns[:len(columns) // 2]).collect()"`

### Issue description

The performance of `pl.LazyFrame.drop()` is drastically reduced since `polars==0.19.2`, i.e. ~5-10x slower. A subsequent call to .collect() is also negatively impacted.

The actual cause is probably in the Rust codepath, possibly introduced in #10885.

### Expected behavior

No significant performance regression vs. `polars==0.19.1`

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.2
Index type:          UInt32
Platform:            macOS-13.5.1-arm64-arm-64bit
Python:              3.11.5 (v3.11.5:cce6ba91b3, Aug 24 2023, 10:50:31) [Clang 13.0.0 (clang-1300.0.29.30)]
----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               1.25.2
pandas:              2.1.0
pyarrow:             <not installed>
pydantic:            2.3.0
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by aploium** [comment]
It should be the extra cost from multi-threading, `select()` will use parallelization, which cause a fixed delay. (see #10316)

for prove, in the benchmark above, if we increase the dataset len from 100 to 10000, they cause almost same time.

```
python -m timeit -s "import polars as pl; columns = tuple(f'column_{i}' for i in range(100)); ldf = pl.LazyFrame({column: tuple(range(10)) for column in columns})" "ldf.drop(columns[:len(columns) // 2])"
# 5000 loops, best of 5: 45.1 usec per loop
```
changes to 1000x data length
```
python -m timeit -s "import polars as pl; columns = tuple(f'column_{i}' for i in range(100)); ldf = pl.LazyFrame({column: tuple(range(10000)) for column in columns})" "ldf.drop(columns[:len(columns) // 2])"
# 5000 loops, best of 5: 45.1 usec per loop
```
they use same time.

So, sadly, I dont think it's possible to fix this without greatly reduce the parallelization's overhead. :(  (I'm at same scenario with you)

**Comment by ritchie46** [comment]
I will take a look. Maybe I can regain some of the performance lost. The benefit of the new `drop` implementation is that it will be triggering projection pushdown and therefor will ensure we don't read those columns from the reader, which will save much more.

**Comment by ritchie46** [comment]
@jakob-keller could you tell me a bit more about how this is so noticable in your environment? I want to understand how this is used.

**Comment by jakob-keller** [comment]
> @jakob-keller could you tell me a bit more about how this is so noticable in your environment? I want to understand how this is used.

We run a service that generates ML inference from streams of high frequency inputs in near-realtime. `polars` is used for two purposes:

1. As inputs / input batches become available, they are ingested into a "primary features" dataframe using `primary_features_df.extend()`.

2. On demand, those "primary features" are transformed into a "final features" vector which is then fed into the ML model. The operation chain is equivalent to `primary_features_df.lazy().with_columns(...).with_columns(...).drop(...).select_seq(...).tail(1).collect(..., disable all optimisations is fastest).row(index=-1)`.

Both code paths are performed dozens to hundreds of times per second and are highly latency sensitive. To guard against performance regressions, we have an internal benchmark that highlights any regressions that might affect our use case.

**Comment by ritchie46** [comment]
Interesting. How is main performing after trying to combat the regression?

**Comment by jakob-keller** [comment]
> Interesting. How is main performing after trying to combat the regression?

I would love to, but `make test` keeps failing when I try to set up my local environment:

```
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.8
🐍 Not using a specific python interpreter
Ignoring pyarrow: markers 'extra == "pyarrow"' don't match your environment
Ignoring pyarrow: markers 'extra == "pandas"' don't match your environment
Ignoring pandas: markers 'extra == "pandas"' don't match your environment
Ignoring numpy: markers 'extra == "numpy"' don't match your environment
Ignoring fsspec: markers 'extra == "fsspec"' don't match your environment
Ignoring connectorx: markers 'extra == "connectorx"' don't match your environment
Ignoring xlsx2csv: markers 'extra == "xlsx2csv"' don't match your environment
Ignoring openpyxl: markers 'extra == "openpyxl"' don't match your environment
Ignoring deltalake: markers 'extra == "deltalake"' don't match your environment
Ignoring backports.zoneinfo: markers 'python_version < "3.9" and extra == "timezone"' don't match your environment
Ignoring tzdata: markers 'platform_system == "Windows" and extra == "timezone"' don't match your environment
Ignoring matplotlib: markers 'extra == "matplotlib"' don't match your environment
Ignoring pydantic: markers 'extra == "pydantic"' don't match your environment
Ignoring sqlalchemy: markers 'extra == "sqlalchemy"' don't match your environment
Ignoring pandas: markers 'extra == "sqlalchemy"' don't match your environment
Ignoring xlsxwriter: markers 'extra == "xlsxwriter"' don't match your environment
Ignoring adbc_driver_sqlite: markers 'extra == "adbc"' don't match your environment
Ignoring cloudpickle: markers 'extra == "cloudpickle"' don't match your environment
Ignoring polars: markers 'extra == "all"' don't match your environment
💻 Using `MACOSX_DEPLOYMENT_TARGET=11.0` for aarch64-apple-darwin by default
   Compiling libz-ng-sys v1.1.12
   Compiling serde_json v1.0.105
   Compiling futures v0.3.28
   Compiling rustc_version v0.4.0
   Compiling brotli v3.3.4
   Compiling lz4 v1.24.0
   Compiling streaming-decompression v0.1.2
   Compiling multiversion-macros v0.7.3
   Compiling bytemuck_derive v1.4.1
   Compiling lock_api v0.4.10
   Compiling parquet-format-safe v0.2.4
   Compiling seq-macro v0.3.5
error: failed to run custom build command for `libz-ng-sys v1.1.12`

Caused by:
  process didn't exit successfully: `/Users/jakobkeller/PycharmProjects/polars/py-polars/target/debug/build/libz-ng-sys-a47180684ab45cd7/build-script-build_zng` (exit status: 101)
  --- stdout
  CMAKE_TOOLCHAIN_FILE_aarch64-apple-darwin = None
  CMAKE_TOOLCHAIN_FILE_aarch64_apple_darwin = None
  HOST_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_aarch64-apple-darwin = None
  CMAKE_GENERATOR_aarch64_apple_darwin = None
  HOST_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_aarch64-apple-darwin = None
  CMAKE_PREFIX_PATH_aarch64_apple_darwin = None
  HOST_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_aarch64-apple-darwin = None
  CMAKE_aarch64_apple_darwin = None
  HOST_CMAKE = None
  CMAKE = None
  running: cd "/Users/jakobkeller/PycharmProjects/polars/py-polars/target/debug/build/libz-ng-sys-eefbc3a11d2e84f6/out/build" && CMAKE_PREFIX_PATH="" "cmake" "/Users/jakobkeller/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libz-ng-sys-1.1.12/src/zlib-ng" "-DCMAKE_OSX_ARCHITECTURES=arm64" "-DBUILD_SHARED_LIBS=OFF" "-DZLIB_COMPAT=OFF" "-DZLIB_ENABLE_TESTS=OFF" "-DWITH_GZFILEOP=ON" "-DCMAKE_INSTALL_PREFIX=/Users/jakobkeller/PycharmProjects/polars/py-polars/target/debug/build/libz-ng-sys-eefbc3a11d2e84f6/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -arch arm64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -arch arm64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -arch arm64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Debug"

  --- stderr
  thread 'main' panicked at /Users/jakobkeller/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:

  failed to execute command: No such file or directory (os error 2)
  is `cmake` not installed?

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `MACOSX_DEPLOYMENT_TARGET="11.0" PYO3_ENVIRONMENT_SIGNATURE="cpython-3.11-64bit" PYO3_PYTHON="/Users/jakobkeller/PycharmProjects/polars/.venv/bin/python" PYTHON_SYS_EXECUTABLE="/Users/jakobkeller/PycharmProjects/polars/.venv/bin/python" "cargo" "rustc" "--message-format" "json-render-diagnostics" "--manifest-path" "/Users/jakobkeller/PycharmProjects/polars/py-polars/Cargo.toml" "--lib" "--" "-C" "link-arg=-undefined" "-C" "link-arg=dynamic_lookup" "-C" "link-args=-Wl,-install_name,@rpath/polars.abi3.so"`
make: *** [build] Error 1
```

**Comment by jakob-keller** [comment]
> Interesting. How is main performing after trying to combat the regression?

I just checked with `polars==0.19.3`: Performance of `collect()` has not returned to the level seen in `polars<=0.19.1`, there might actually be a small, additional performance degradation. 🐌 Overall, the impact of recent improvements elsewhere more than compensates, though.

I have been able to eliminate the use of `drop()` (and the preceding `with_columns()`) by switching to a `select()` operation. 🚀
